doc/rados/operations/stretch-mode.rst

```diff
@@ -48,26 +48,39 @@ place that will cause the cluster to re-replicate the data until the
-you can restore service, Ceph will return to normal functionality on its own.
-
-But there are scenarios where you lose data availability despite having
-enough servers available to satisfy Ceph's consistency and sizing constraints, or
-where you may be surprised to not satisfy Ceph's constraints.
-The first important category of these failures resolve around inconsistent
-networks -- if there's a netsplit, Ceph may be unable to mark OSDs down and kick
-them out of the acting PG sets despite the primary being unable to replicate data.
-If this happens, IO will not be permitted, because Ceph can't satisfy its durability
-guarantees.
-
-The second important category of failures is when you think you have data replicated
-across data centers, but the constraints aren't sufficient to guarantee this.
-For instance, you might have data centers A and B, and your CRUSH rule targets 3 copies
-and places a copy in each data center with a ``min_size`` of 2. The PG may go active with
-2 copies in site A and no copies in site B, which means that if you then lose site A you
-have lost data and Ceph can't operate on it. This situation is surprisingly difficult
-to avoid with standard CRUSH rules.
+
+Ceph does not permit the compromise of data integrity and data consistency
+under any circumstances. When service is restored after a network failure of a
+loss of Ceph nodes, Ceph will restore itself to a state of normal functioning
+without operator intervention.  
+
+Ceph does not permit the compromise of data integrity or data consistency, but
+there are situations in which data availability is compromised. These
+situations can occur even though there are enough clusters available to satisfy
+Ceph's consistency and sizing constraints. In some situations, you might
+discover that your cluster does not satisfy those constraints.
+
+The first category of these failures that we will discuss involves inconsistent
+networks -- if there is a netsplit (a disconnection between two servers that
+splits the network into two pieces), Ceph might be unable to mark OSDs ``down``
+and remove them from the acting PG sets. This failure to mark ODSs ``down``
+will occur, despite the fact that the primary PG is unable to replicate data (a
+situation that, under normal non-netsplit circumstances, would result in the
+marking of affected OSDs as ``down`` and their removal from the PG). If this
+happens, Ceph will be unable to satisfy its durability guarantees and
+consequently IO will not be permitted.
+
+The second category of failures that we will discuss involves the situation in
+which the constraints are not sufficient to guarantee the replication of data
+across data centers, though it might seem that the data is correctly replicated
+across data centers. For example, in a scenario in which there are two data
+centers named Data Center A and Data Center B, and the CRUSH rule targets three
+replicas and places a replica in each data center with a ``min_size`` of ``2``,
```

Comment:
@gregsfortytwo: I think that this part of the text needs to have more explanation of why a CRUSH rule that "targets three replicas and places a replica in each data center with a min_size of 2" might result in a PG that is messed up. In this connection I offer my own confusion (and ignorance) in order to give the discussion focus. I think I need more understanding of what "targets three replicas" means.

Reply:
Maybe a better phrasing is "selects three replicas, with at least one replica in each data center and min_size 2".

This paragraph actually looks pretty good to me.

Reply:
> Maybe a better phrasing is "selects three replicas, with at least one replica in each data center and min_size 2".
> 
> This paragraph actually looks pretty good to me.

I'm pleased that you think the paragraph looks good, but I'm still confused by the text in it. What does it mean for a CRUSH rule to select three replicas with at least one replica in each data center and min_size 2? It seems to me that what we're trying to do with such a CRUSH rule is to create a situation in which Data Center A is targeted and also Data Center B is targeted and also a third, external, data center is targeted. But I could be totally wrong, and I don't think the text makes it sufficiently unambiguous.

I apologize in advance for what might turn into a very frustrating exchange, but I promise that I am not affecting obtuseness in order to be captious.

Reply:
Okay, so min_size as I referred to in this changed text is the number of replicas required for a PG to go active. But that's not a crush constraint at all, and I suspect maybe you're looking down at the CRUSH rule at the bottom of the document where it says "min_size 1". Which is understandable but is actually a *completely different* concept.

The purpose of CRUSH is to select OSDs to host a given PG; we often call each of these a replica. You can generally use the same CRUSH rule to select a wide range of OSDs, so the rule lets you specify a valid range (from min_size to max_size). We mostly just say 1 and 10, which is what I did here — but that is actually a lie. What that rule really does is "take two OSDs from site1, and then take 2 OSDs from site2".

If you used this rule but set the pg size to 4 instead of 3, CRUSH would generate 4 OSDs while evaluating the rule, but then only return the first 3 of them. So every PG would have two replicas in site1 and one replica in site2.

At that point, if you rebooted a node in site2 while continuing to operate the cluster, some PGs would only have replicas in site1 — but they would have two replicas in that one site, and you are allowed to go active with two replicas, so they would. If you subsequently lost site1 before all the OSDs in site2 came back and were updated, you would lose data.